x/fswap/keeper/keeper.go

```diff
@@ -1,42 +1,345 @@
 package keeper
 
 import (
+	"errors"
 	"fmt"
 
 	"github.com/tendermint/tendermint/libs/log"
 
 	"github.com/Finschia/finschia-sdk/codec"
+	"github.com/Finschia/finschia-sdk/store/prefix"
 	storetypes "github.com/Finschia/finschia-sdk/store/types"
 	sdk "github.com/Finschia/finschia-sdk/types"
+	sdkerrors "github.com/Finschia/finschia-sdk/types/errors"
 	"github.com/Finschia/finschia-sdk/x/fswap/types"
 )
 
-type (
-	Keeper struct {
-		cdc           codec.BinaryCodec
-		storeKey      storetypes.StoreKey
-		accountKeeper types.AccountKeeper
-		bankKeeper    types.BankKeeper
-		config        types.Config
-	}
-)
+type Keeper struct {
+	cdc      codec.BinaryCodec
+	storeKey storetypes.StoreKey
 
-func NewKeeper(
-	cdc codec.BinaryCodec,
-	storeKey storetypes.StoreKey,
-	ak types.AccountKeeper,
-	bk types.BankKeeper,
-	config types.Config,
-) Keeper {
+	AccountKeeper
+	BankKeeper
+
+	fromDenom    string
+	toDenom      string
+	swapMultiple sdk.Int
+	swapCap      sdk.Int
+}
+
+func NewKeeper(cdc codec.BinaryCodec, storeKey storetypes.StoreKey, ak AccountKeeper, bk BankKeeper) Keeper {
 	return Keeper{
 		cdc:           cdc,
 		storeKey:      storeKey,
-		accountKeeper: ak,
-		bankKeeper:    bk,
-		config:        config,
+		AccountKeeper: ak,
+		BankKeeper:    bk,
+		fromDenom:     "",
+		toDenom:       "",
+		swapMultiple:  sdk.ZeroInt(),
+		swapCap:       sdk.ZeroInt(),
 	}
 }
 
 func (k Keeper) Logger(ctx sdk.Context) log.Logger {
 	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
 }
+
+func (k Keeper) FswapInit(ctx sdk.Context, fswapInit types.FswapInit) error {
+	if err := fswapInit.ValidateBasic(); err != nil {
+		return err
+	}
+	if k.hasBeenInitialized(ctx) {
+		return errors.New("already initialized")
+	}
+	if err := k.setFswapInit(ctx, fswapInit); err != nil {
+		return err
+	}
+	swapped := types.Swapped{
+		OldCoinAmount: sdk.Coin{
+			Denom:  fswapInit.GetFromDenom(),
+			Amount: sdk.ZeroInt(),
+		},
+		NewCoinAmount: sdk.Coin{
+			Denom:  fswapInit.GetToDenom(),
+			Amount: sdk.ZeroInt(),
+		},
+	}
+	if err := k.setSwapped(ctx, swapped); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k Keeper) Swap(ctx sdk.Context, addr sdk.AccAddress, oldCoinAmount sdk.Coin) error {
+	if ok := k.HasBalance(ctx, addr, oldCoinAmount); !ok {
+		return sdkerrors.ErrInsufficientFunds
+	}
+	fswapInit, err := k.getFswapInit(ctx)
+	if err != nil {
+		return err
+	}
+	if oldCoinAmount.GetDenom() != fswapInit.GetFromDenom() {
+		return errors.New("denom mismatch")
+	}
+
+	newAmount := oldCoinAmount.Amount.Mul(fswapInit.SwapMultiple)
```

Comment:
You may check the restriction after the swap has been completed (or in `updateSwapped()`).

Reply:
There are some updated code.
Please check updated code has the same problem.